cwms_radar_api/src/main/java/cwms/radar/security/KeyAccessManager.java

```diff
@@ -0,0 +1,195 @@
+        }		
+
+	}
+
+    /**
+     * This isn't currently used from Tomcat so we don't have any issues with the 
+     * session; just use the connection itself.
+     * 
+     * NOTE: we will need to expand a bit in the future, but I think another JAPSIC or valve
+     * handler that sets up the principal for the CWMS Access Manager will likely
+     * have easier workflow.
+     * @param ctx
+     * @param user
+     * @param key the key
+     */
+    private void prepareContextWithUser(Context ctx, String user,String key) throws SQLException {
+        logger.info("Validated Api Key for user=" + user);
+
+        ConnectionPreparer keyPreparer = new ApiKeyUserPreparer(key);
+        ConnectionPreparer officePrepare = new SessionOfficePreparer(ctx.queryParam("office"));
+        //ConnectionPreparer resetPreparer = new DirectUserPreparer("q0webtest");
+        DelegatingConnectionPreparer apiPreparer = new DelegatingConnectionPreparer(officePrepare,keyPreparer);
+
+        if (dataSource instanceof ConnectionPreparingDataSource) {
+            ConnectionPreparingDataSource cpDs = (ConnectionPreparingDataSource)    dataSource;
+            ConnectionPreparer existingPreparer = cpDs.getPreparer();
+
+            // Have it do our extra step last.
+            cpDs.setPreparer(new DelegatingConnectionPreparer(existingPreparer, apiPreparer));
+        } else {            
+            ctx.attribute(ApiServlet.DATA_SOURCE, new ConnectionPreparingDataSource(apiPreparer, dataSource));
+        }
+    }
+
+    private String authorized(Context ctx, String key, Set<RouteRole> routeRoles)
+    {
+        String retval = null;
+        String user = checkKey(key,ctx);
+
+        if (routeRoles == null || routeRoles.isEmpty()) {
+            retval = user;
+        } else {
+            Set<RouteRole> specifiedRoles = getRoles(user,ctx.queryParam("office"));
+            retval = specifiedRoles.containsAll(routeRoles) == true ? user : null;
+        }
+
+        return retval;
+    }
+
+    private Set<RouteRole> getRoles(String user,String office) {
+        Set<RouteRole> roles = new HashSet<>();
+        try(Connection conn = dataSource.getConnection();
+            PreparedStatement getRoles = conn.prepareStatement(RETRIEVE_GROUPS_OF_USER);
+        ) {
+            getRoles.setString(1,user);
+            getRoles.setString(2,office);
+            try (ResultSet rs = getRoles.executeQuery()) {
+                while(rs.next()) {
+                    roles.add(new Role(rs.getString(1)));
+                }
+            }
+        } catch(SQLException ex) {
+            logger.log(Level.WARNING,"Failed to retrieve roles for user",ex);
+        }
+        return roles;
+    }
+
+    private String checkKey(String key, Context ctx) {
+        try(Connection conn = dataSource.getConnection();
+            PreparedStatement setApiUser = conn.prepareStatement("begin cwms_env.set_session_user_direct(upper('q0hectest_pu')); end;");
```

Comment:
q0hectest_pu seems like it should be pulled out into a constant. Is that a new user for auth checks?

Reply:
oy, thanks, that actually needs to be a "select user from dual" call.